rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,387 @@
+This is similar to other specification bugs.
+All the stubbing code will be available, so it is possible to inspect the assumptions it makes.
+
+### Comparison to function contracts
+
+- The [currently proposed function contract mechanism](https://github.com/model-checking/kani/tree/features/function-contracts) does not provide a way to specify contracts on external functions.
+This is one of the key motivations for stubbing.
+- In many cases, stubs are more user-friendly than contracts.
+With contracts, it is sometimes necessary to explicitly provide information that is automatically captured in Rust (such as which memory is written).
+Furthermore, contract predicates constitute a DSL of their own that needs to be learned; using stubbing, we can stick to using just Rust.
+
+### Alternative #1: Annotate stubbed functions
+
+In this alternative, users add an attribute `#[kani::stub_by(<replacement>)]` to the function that should be replaced.
+This approach is similar to annotating a function with a contract specifying its behavior (the stub acts like a programmatic contract).
+The major downside with this approach is that it would not be possible to stub external code. We see this as a likely use case that needs to be supported: users will want to replace `std` library functions or functions from arbitrary external crates.
+
+### Alternative #2: Annotate stubs
+
+In this alternative, users add an attribute `#[kani::stub(<original>)]` to the stub function itself, saying which function it replaces:
+
+```rust
+#[cfg(kani)]
+#[kani::stub("rand::random")]
+fn mock_random<T: kani::Arbitrary>() -> T { ... }
+```
+
+The downside is that this stub must be uniformly applied across all harnesses and the stub specifications might be spread out across multiple files.
+It would also require an extra layer of indirection to use a function as a stub if the user does not have source code access to it.
+
+### Alternative #3: Annotate harnesses and stubs 
+
+This alternative combines the proposed solution and Alternative #2.
+Users annotate the stub (as in Alternative #2) and specify for each harness which stubs to use using an annotation `#[kani::use_stubs(<stub>+)]` placed above the harness.
+
+This could be combined with modules, so that a module can be used to group stubs together, and then harnesses could pull in all the stubs in the module:
+
+```rust
+#[cfg(kani)]
+mod my_stubs {
+
+  #[kani::stub("foo")]
+  fn stub1() { ... }
+
+  #[kani::stub("bar")]
+  fn stub2() { ... }
+
+}
+
+#(cfg[kani])
+#[kani::proof]
+#[kani::use_stubs("my_stubs")]
+fn my_harness() { ... }
+```
+
+The benefit is that stubs are specified per harness, and (using modules) it might be possible to group stubs together.
+The downside is that multiple annotations are required and the stub mappings themselves are remote from the harness.
```

Comment:
I don't understand the downside. We don't lose the ability to still "cherry-pick" things from a module right? Stub-sets are just as "remote" as a module like this, no?

It seems like this might be a "nicer" and more Rust-friendly way of writing stub-sets and not actually have any real downsides that I can see...

Reply:
Some stronger downsides with this approach:

1. How do you mock multiple functions with the same stub? (Especially if, say, harness A wants to use `stub1` to mock `foo`, and harness B wants to use `stub1` to mock `bar`.)
2. How do you combine stub sets defined via modules? (This is probably possible, but would need to be fleshed out.)
3. Is it a regular module or not? Modules can contain other things than functions - how should we interpret them?